tensorflow/python/ops/nn_ops.py

```diff
@@ -1626,6 +1626,10 @@ def _swap_axis(logits, dim_index, last_index, name=None):
   # If dim is not the last dimension, we have to do a reshape and transpose so
   # that we can still perform softmax on its last dimension.
 
+  # In case dim is negative (and is not last dimension -1), add shape.ndims
+  if dim < 0:
+    dim += shape.ndims
```

Comment:
Could `dim`  be invalid negative value, say `-5 += 3`?

Reply:
@facaiy In case dim is invalid negative `-5` then a `ValueError` will be thrown.

Reply:
Yup, however `-4 + 3 = -1` would pass, right? Perhaps it's better to check the argument here and raise an invalid value exception early?